### PR TITLE
[1] Upgrade build process for plugins

### PR DESCRIPTION
**Pull request by MatApple**
## Merging into feature branch
- Remove hard coded plugins reference. Now set external plugin location with env variable.
- Generate jest config dynamically.

NOTE: Tests will fail till we merge into feature branch

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @Poltergeist

File: gulpfile.js [hunk 1]
```diff
@@ -8,83 +8,40 @@ if (!fs.existsSync(configFilePath)) {
 var autoprefixer = require('gulp-autoprefixer');
 var browserSync = require('browser-sync');
 var colorLighten = require('less-color-lighten');
-var changed = require('gulp-changed');
+// var changed = require('gulp-changed');
```
Comment (rcorral) [inline gulpfile.js]: Do you uncomment these on a later PR or why are they uncommented here?

File: package.json [hunk 2]
```diff
@@ -74,18 +81,28 @@
...
     "mkdirp": "0.5.1",
     "node-libs-browser": "1.0.0",
+    "postcss-loader": "0.8.2",
     "run-sequence": "1.1.5",
     "source-map-loader": "0.1.5",
+    "string-replace-webpack-plugin": "0.0.3",
+    "style-loader": "0.13.1",
     "transform-loader": "0.2.3",
     "webpack": "1.13.0",
+    "webpack-dev-server": "1.14.1",
```
Comment (rcorral) [inline package.json]: Oh maaan, so many new dependencies :( I hope we're removing some other ones to offset.

File: package.json [hunk 3]
```diff
@@ -47,21 +48,27 @@
...
     "babel-loader": "6.2.4",
     "babel-plugin-transform-runtime": "6.7.5",
     "babel-preset-es2015": "6.6.0",
+    "babel-preset-jest": "11.0.2",
     "babel-preset-react": "6.5.0",
+    "babel-preset-react-hmre": "1.1.1",
     "babel-runtime": "6.6.1",
     "browser-sync": "2.12.3",
+    "copy-webpack-plugin": "2.1.3",
+    "cross-env": "1.0.7",
```
Comment (rcorral) [inline package.json]: Do we need this?

File: package.json [hunk 4]
```diff
@@ -47,21 +48,27 @@
...
-    "babel-jest": "11.0.2",
     "babel-loader": "6.2.4",
     "babel-plugin-transform-runtime": "6.7.5",
     "babel-preset-es2015": "6.6.0",
+    "babel-preset-jest": "11.0.2",
     "babel-preset-react": "6.5.0",
+    "babel-preset-react-hmre": "1.1.1",
     "babel-runtime": "6.6.1",
     "browser-sync": "2.12.3",
+    "copy-webpack-plugin": "2.1.3",
```
Comment (rcorral) [inline package.json]: When do we use this?

File: package.json [hunk 5]
```diff
@@ -27,6 +27,7 @@
     "clipboard": "1.5.10",
     "cookie": "0.2.3",
     "d3": "3.5.16",
+    "file-loader": "0.8.5",
```
Comment (rcorral) [inline package.json]: Should this be a dev dependency?
and where does it get used?

File: package.json [hunk 6]
```diff
@@ -74,18 +81,28 @@
     "gulp-sourcemaps": "1.6.0",
     "gulp-uglify": "1.5.3",
     "gulp-util": "3.0.7",
+    "handlebars": "4.0.5",
```
Comment (rcorral) [inline package.json]: wat? do we need this for?

File: jest/preprocessor.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
-var babelJest = require('babel-jest');
 var webpackAlias = require('jest-webpack-alias');
+var babel = require('babel-core');
```
Comment (mlunoe) [inline jest/preprocessor.js]: 🔤

File: package.json [hunk 8]
```diff
@@ -94,74 +111,24 @@
...
-    "testing": "NODE_ENV='testing' ./node_modules/.bin/gulp serve",
+    "testing": "cross-env NODE_ENV='testing' ./node_modules/.bin/gulp serve",
+    "lint-plugins": "PLUGINS=$DCOS_UI_PLUGINS; ./node_modules/.bin/eslint \"$PLUGINS/**/*.js\" --quiet",
     "lint": "./node_modules/.bin/eslint '**/*.js' --quiet",
-    "livereload": "NODE_ENV='development' ./node_modules/.bin/gulp livereload",
+    "livereload": "cross-env NODE_ENV='development' ./node_modules/.bin/gulp livereload",
     "livereload-notify": "export NOTIFY=true && npm run livereload",
-    "test": "./node_modules/.bin/gulp copy:external-plugins && npm run lint && jest"
+    "test": "npm run lint && npm run lint-plugins && node jest/genConfig.js && NODE_PATH=node_modules jest --config=jest/config.json --no-cache",
+    "jest": "node jest/genConfig.js && NODE_PATH=node_modules jest --config=jest/config.json --no-cache",
```
Comment (mlunoe) [inline package.json]: No `cross-env` here?

Comment (mlunoe) [inline package.json]: Not sure what problem the `cross-env` actually solves

Comment (MatApple) [inline package.json]: @mlunoe Our npm scripts will choke on Windows. However, I think everyone seems to be in favor of only providing support for build requirements as they are needed. - From `cross-env` image

Comment (mlunoe) [inline package.json]: Arh! Thank you! I was looking for that explanation!
Ok

File: gulpfile.js (hunk 1 above)
Comment (MatApple) [inline gulpfile.js]: Fixed

File: package.json (hunk 2 above)
Comment (MatApple) [inline package.json]: Yes, it has balanced out :)

File: package.json (hunk 3 above)
Comment (MatApple) [inline package.json]: Removed until there is a stronger case for having it.

File: package.json (hunk 4 above)
Comment (MatApple) [inline package.json]: Removed. Was used to copy Dygraphs but we might bundle Dygraphs differently.

File: package.json (hunk 5 above)
Comment (MatApple) [inline package.json]: Dev dependency. Used to load images.

File: package.json (hunk 6 above)
Comment (MatApple) [inline package.json]: removed

File: jest/preprocessor.js (hunk 7 above)
Comment (MatApple) [inline jest/preprocessor.js]: done

File: package.json (hunk 8 above)
Comment (MatApple) [inline package.json]: Removed all `cross-env` till there is a stronger case for using it.

**Comment by MatApple** [comment]
@rcorral @mlunoe - I updated based on your comments. Also committed new shrinkwrap after cleaning dependencies.

**Comment by mlunoe** [comment]
Needs a rebase, but then 🚢!

**Pull request merged**